cornell-dti/course-plan
# Two-graph individual requirement progress and warnings

### Summary 

This pull request implements two-graph progress tracking for individual requirements based off both @SamChou19815 's experimental code #599, and warnings on individual requirement slots and courses where conflicts occur based on Yuxuan's designs [here](https://www.figma.com/file/vz9T397rHdewbB2McmUDEr/Yuxuan-Working-File?node-id=2035%3A1560)

- [x] Add warnings to requirement bar and schedule view on courses that have conflicts (by updating `CourseCaution`)
- [x] Color courses in req bar and plan when they have these conflicts
- [x] Refactor code to handle two-graphs and display number of combined safe/dangerous courses fulfilling individual requirement slots in their progress counters (including compound requirements)
- [x] **NEW** Use  `RequirementConflictsEnabled` GK so only those with GK enabled see changes.



Remaining TODOs:

- [x] Add dummy data or a gatekeeper in some way so this can be better tested
- [x] In the next PR, color progress bar yellow and add new warning when requirements within it have dangerous fulfillments (refer to Sam's PR)
- [ ] In a PR afterwards, add the modal to make requirements choices, and link to the `FIX NOW` hyperlink in the warnings.

### Test Plan 


To test the below, I updated the function `buildRequirementFulfillmentGraphFromUserData` in `requirement-graph-builder-from-user-data.ts` locally with a harcoded `doubleCountedCourseUniqueIDSet` of `uniqueIds` corresponding to a few courses in my plan (which can be found on Firestore) as there is no way to add conflicts as of now.

![image](https://user-images.githubusercontent.com/25535093/146719388-c684f11f-eaa7-4a7b-9f44-eb59a58c1577.png)

1. Confirm requirements with conflicts show the total number of fulfillments, even if dangerous, and courses that conflict have warnings and yellow highlight underneath them.
![image](https://user-images.githubusercontent.com/25535093/146718487-6c37f86d-6540-4550-8fa8-53eee25457a0.png)
![image](https://user-images.githubusercontent.com/25535093/146718514-e398303d-45db-4748-a625-2e5ab683533a.png)

2. Confirm courses with conflicts have warnings and yellow highlight in semester plan (and conflict warning comes first if there are multiple).

![image](https://user-images.githubusercontent.com/25535093/146718373-e8bb7cb7-3672-4837-819f-cf7045b09442.png)
![image](https://user-images.githubusercontent.com/25535093/146718359-8eb15b0a-8eb7-4496-b1fa-21305905ce87.png)

3. Confirm fulfillment counting works correctly for compounded requirements and credit requirements. Specifically, that they count dangerous fulfillments and can be completed.
![image](https://user-images.githubusercontent.com/25535093/146718896-db51671f-b734-4456-a724-2d9799d6df39.png)
![image](https://user-images.githubusercontent.com/25535093/146718964-4a5cbe8b-0da9-45c7-9af0-3dff1e8ddb14.png)

4. Confirm other warning functionality is unchanged.

5. **NEW** Confirm that disabling the GK turns off all the above changes. 
![image](https://user-images.githubusercontent.com/25535093/160694921-154f3645-a249-44ae-9a35-655e635deed6.png)
![image](https://user-images.githubusercontent.com/25535093/160694876-503afaba-949a-448f-9b26-7dd8de83e630.png)

### Notes 

This can be safely merged with no changes to the user - ~~as there is no way for users to have any conflicts yet to my knowledge.~~ as there is now a GK

Design questions remaining...
1. If a requirement has conflicts that complete it, should the req show up in "completed requirements" ?
2. If a course has a yellow highlight and is clicked on in the semester plan, should the blue or yellow border show? Right now I give the blue prioritization.
